foundations/src/settings/mod.rs

```diff
@@ -461,24 +485,117 @@ pub fn to_yaml_file(settings: &impl Settings, path: impl AsRef<Path>) -> Bootstr
     )?)
 }
 
-/// Parse settings from YAML string.
+/// Builds [SettingsDeserializationOptions] with specified options.
+/// Any left unspecified will use defaults.
+pub struct SettingsDeserializationOptionsBuilder {
+    ignored_key_callback: Option<Box<dyn FnMut(IgnoredPath)>>,
+}
+
+impl SettingsDeserializationOptionsBuilder {
+    /// Initialize the builder without any options specified
+    pub fn new() -> Self {
+        SettingsDeserializationOptionsBuilder {
+            ignored_key_callback: None,
+        }
+    }
+
+    /// Set the callback that's called when YAML keys are not used during deserialization
+    /// into settings.
+    ///
+    /// If unspecified, the default will log each key as a warning, as long as it does not
+    /// end with [YAML_KEY_ALLOW_UNUSED_SUFFIX]. In that case, nothing will be logged.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use foundations::settings::{IgnoredPath, SettingsDeserializationOptionsBuilder};
+    ///
+    /// let builder = SettingsDeserializationOptionsBuilder::new();
+    /// builder.with_ignored_key_callback(Box::new(|path| {
+    ///     // Do something
+    /// }));
+    ///
+    /// ```
+    pub fn with_ignored_key_callback(mut self, callback: Box<dyn FnMut(IgnoredPath)>) -> Self {
+        self.ignored_key_callback = Some(callback);
+        self
+    }
+
+    /// Creates the configured [SettingsDeserializationOptions]
+    ///
+    /// Any values left unspecified will use default values.
+    pub fn build(self) -> SettingsDeserializationOptions {
+        SettingsDeserializationOptions {
+            ignored_key_callback: self.ignored_key_callback.unwrap_or_else(|| {
```

Comment:
This is changing the default behavior to log.  Should the default be to do nothing, and let users opt-in?

Reply:
There are two contexts here - the context of just this builder and the whole context of the default behavior of the `from_yaml_str` method (which is eventually called by `Cli::new`). I'm not clear which you're referring to. Though maybe it doesn't matter, as I think the default should match in both cases.

I believe the chain through `Cli::new` is the main way this method is invoked - for example, that's how Oxy gets its settings initially. Both opt-in and opt-out are difficult in this context. I didn't see a clean way to add an option to that top-level method (and pass it down) without either a breaking API change or a non-breaking change by doing something similar to what I did with the builder + new method. A breaking API change for such a small bit of functionality didn't seem right to me, and I'm hesitant to add to the top-level API in `Cli` either.

As far as whether it should be the default - that's a philosophical choice about the design of foundations that I could go either way on. But I lean towards it being on by default since it could be helpful and makes foundations a bit more of a 'batteries included' kind of library. There is also currently the (clunky) way to opt-out using the `YAML_KEY_ALLOW_UNUSED_SUFFIX` in the yaml itself.